### PR TITLE
Implement cache testing

### PR DESCRIPTION
**Pull request by Jackbaude**
Changed cache to no longer take a file and instead take a writer.

Allows for testing of caching module inside of memory. Finsihed testing for add_lease, remove_lease, and update_lease

Signed-off-by: Jack <jackbaude@gmail.com>

### CODE REVIEW

File: src/server.rs [hunk 1]
```diff
@@ -207,7 +206,18 @@ pub async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
 
-    let cache = match LeaseCache::new(conf_dir) {
+    let fq_path = Path::new(DEFAULT_CONFIG_DIR).join("nv-leases");
+    let file = match File::create(&fq_path) {
+        Ok(file) => {
+            debug!("Successfully created leases file: {:?}", fq_path);
+            file
+        }
+        Err(e) => {
+            panic!("Error creating leases file: {:?}", e);
```
Comment (baude) [inline src/server.rs:216]: we cannot have a panic like this

File: src/cache.rs [hunk 2]
```diff
@@ -86,97 +102,335 @@ impl LeaseCache {
...
-    fn default() -> Self {
-        LeaseCache {
-            mem: HashMap::<String, Vec<NetavarkLease>>::new(),
-            path: PathBuf::from(DEFAULT_CACHE_DIR),
+        let writer = &mut self.writer;
+        // Clear the writer so we can add the old leases
+        match writer.clear() {
+            Ok(_) => {}
+            Err(e) => {
+                error!("{:?}", e);
```
Comment (baude) [inline src/cache.rs:160]: make error more descriptive

File: src/cache.rs [hunk 3]
```diff
@@ -86,97 +102,335 @@ impl LeaseCache {
...
-    fn update() {
-        // 1. Clean the directory to the lease
-        // 2. Create a new cache instance
-        // 3. Check that the file to the cache exists
-        // 4. Change the value of that lease and call the update method
-        // 5. Check that the old lease doesnt exist and the new lease is up to date
-        // 6. Clean the directory to the lease
+    fn add_leases() {
+        // Instead of using a file use a cursor
+        let buff = Cursor::new(Vec::new());
```
Comment (baude) [inline src/cache.rs:222]: wwyt about putting most this is setup function so you can re-use in more tests?

File: src/cache.rs [hunk 4]
```diff
@@ -86,97 +102,335 @@ impl LeaseCache {
...
+            // Find the lease in the set of deserialized leases
+            let deserialized_lease = s
+                .get(macaddr)
+                .expect("Could not get the mac address from the map")
+                .get(0)
+                .expect("Could not get lease from set of mac addresses")
+                .clone();
+            // Assure that the amount of leases added is correct amount
+            assert_eq!(s.len(), (i + 1) as usize);
+            // Assure that the lease added was correct
```
Comment (baude) [inline src/cache.rs:265]: i'd like to see you add a test for two identical leases

File: src/cache.rs [hunk 5]
```diff
@@ -86,97 +102,335 @@ impl LeaseCache {
...
+                .unwrap_or_else(|_| panic!("Could not remove {:?} from leases", macaddr));
+            // Assure the lease is no longer in memory
+            assert_eq!(deserialized_lease, removed_lease);
+            assert_eq!(s.len(), (range - i) as usize);
+
+            // Deserialize the cache again to assure the lease is not in the writer
+            let lease_bytes = cache.writer.get_ref().as_slice();
+            let s: HashMap<String, Vec<NetavarkLease>> = match serde_json::from_slice(lease_bytes) {
+                Ok(s) => s,
+                Err(e) => panic!("Error: {:?}", e),
```
Comment (baude) [inline src/cache.rs:349]: we cannot panic like this

File: src/cache.rs [hunk 6]
```diff
@@ -86,97 +102,335 @@ impl LeaseCache {
...
+            assert_eq!(s.len(), (i + 1) as usize);
+            // Assure that the lease added was correct
+            assert_eq!(lease, deserialized_lease);
+        }
+        for i in 0..range {
+            // Deserialize the written bytes to compare
+            let lease_bytes = cache.writer.get_ref().as_slice();
+            let s: HashMap<String, Vec<NetavarkLease>> = match serde_json::from_slice(lease_bytes) {
+                Ok(s) => s,
+                Err(e) => panic!("Error: {:?}", e),
```
Comment (baude) [inline src/cache.rs:324]: no panic

File: src/cache.rs [hunk 7]
```diff
@@ -86,97 +102,335 @@ impl LeaseCache {
...
+                .get(macaddr)
+                .expect("Could not get the mac address from the map")
+                .get(0)
+                .expect("Could not get lease from set of mac addresses")
+                .clone();
+
+            let removed_lease = cache
+                .remove_lease(macaddr)
+                .unwrap_or_else(|_| panic!("Could not remove {:?} from leases", macaddr));
+            // Assure the lease is no longer in memory
```
Comment (baude) [inline src/cache.rs:341]: re-use setup code

File: src/cache.rs [hunk 8]
```diff
@@ -86,97 +102,335 @@ impl LeaseCache {
...
+            assert_eq!(deserialized_lease, removed_lease);
+            assert_eq!(s.len(), (range - i) as usize);
+
+            // Deserialize the cache again to assure the lease is not in the writer
+            let lease_bytes = cache.writer.get_ref().as_slice();
+            let s: HashMap<String, Vec<NetavarkLease>> = match serde_json::from_slice(lease_bytes) {
+                Ok(s) => s,
```
Comment (baude) [inline src/cache.rs:348]: write a test for removing a lease that does not exist.

**Comment by baude** [comment]
@Jackbaude really on the right track ... couple of general comments:

1. i didnt realize how much of my previous commit you had brought in.  let's catch up and get it removed.  if i get time, i will fork your branch and give it a whirl.
2. i would like to see more variety of unit tests including forcing failures and ensuring errors occur.
3. ideally we could find a way to test your lock approach by flooding it.  a good mix of adding/removing would be also ideal especially if randomized.
4. in reviewing your tests,  i wonder if we should worry about the JSON file being corrupted BUT the service has not gone down (like out of disk space).  it would be nice to able to cope with that.  we couild just add that as an issue.

great work

**Comment by baude** [comment]
please rebase and consider review comments ...

File: src/cache.rs (hunk 4 above)
Comment (Jackbaude) [reply src/cache.rs:265]: Can you explain what you mean by this comment?

Comment (baude) [reply src/cache.rs:295]: i dont see the issue you were supposed to record.

**Comment by baude** [comment]
take care of the issue and it is ready i think

**Pull request merged**